### ISSUE TITLE
Urls with . in them are assumed to be files

### ISSUE THREAD

**Issue by wolfadex** [dillonkearns/elm-pages#209]
When rendering routes, `elm-pages` [checks to see if the path is to a file](https://github.com/dillonkearns/elm-pages/blob/b78f8c008b947f5eec665c6b579ef765de3b13b5/src/RenderRequest.elm#L116). It does this by checking for any `.` in the path. When given a URL such as `/wolfadex/tiler/1.0.2` the [isFile](https://github.com/dillonkearns/elm-pages/blob/b78f8c008b947f5eec665c6b579ef765de3b13b5/src/RenderRequest.elm#L143) function assumes that `1.0.2` is a file name and treats it as such. This causes it to 404 instead of navigating to the page for `/:author/:project/:version`.

Part of the complication with this is trying to differentiate a page like `Page/About.elm` and an api like `/about`.

**Comment by sylbru** [comment]
What about something like checking that the last part of the path contains a single dot, followed by something that looks like an extension (like `[^.]+\.[a-b][a-b0-9]{1,3}`). That would definitely get false positives (also some false negatives I guess), but less than currently. Still, a half solution more than anything :)

**Comment by dillonkearns** [comment]
Yeah, somehow it seems like there's a problem in the overall architecture here that needs to be revisited somehow.

I guess the challenge is that I'm trying to have a deterministic kind of routing where it's clear where a route will be handled without digging into any specific code for listing out exceptions or lists of values to handle.

I've achieved that pretty well within the scope of the UI routes - if you have `app/Route/About/Team.elm` and `app/Route/About/Name_.elm`, then you know that `/about/team` will be handled by the more specific module, and anything else will be handled by the more general one. The SPLAT routing makes this a little more subtle because you can have catchall handling, but it's the same basic idea that the more specific a route the higher its precedence (more general ones will be overshadowed). I find that pretty intuitive and I'm happy with it.

But now the API routes become pretty subtle when you add it to the mix. Originally I was thinking about checking for browser navigation requests, but I ended up finding it to be a bit flaky because it feels like the environment (whether you make a request from a browser or curl, etc.) changes the response you get back, which feels strange. Also this doesn't neatly solve the issue when it comes to generating a file on the file path in a static build.

A lot of frameworks, such as NextJS, simplify this model by having a specific namespace for API routes, `/api/...`, that's reserved for these API routes. This feels pretty limiting, though.

I think I'm still searching for a model where the API routing precedence fits in with the UI routes, but at a lower precedence (so any UI routes will overshadow an API route at the same level of specificity). But the problem there is that sometimes you want to have UI routes and API routes with the same general pattern, but just having a file extension for the API route version.

What if paths that could match either a UI route or an API route would go to the matching API route if they contain a dot character, would that be a reasonable heuristic? Does that fit neatly with that mental model that allows you to think about it cleanly? It's at least one possible direction to consider.
